muscan/muscan.1
.TH "MUSCAN" "1" "Release 0.2.0" "daelstorm" "Museek Daemon Plus"
.SH "NAME"
.LP 
\fBMuscan\fR \- File scanner for Museek
.SH "SYNOPSIS"
.B muscan
[\-c <\fIfilename\fP>] [\-\-config <\fIfilename\fP>]
[\-b] [\-\-buddy]
[\-v] [\-\-verbose]
[\-r] [\-\-rescan]
[\-n] [\-\-noscan]
[\-s <\fIdirectory\fP>] [\-\-share <\fIdirectory\fP>]
[\-u <\fIdirectory\fP>] [\-\-unshare <\fIdirectory\fP>]
[\-l] [\-\-list]
[\-h] [\-\-help]
.SH "DESCRIPTION"
.LP 
Muscan scans paths & files to be shared by \fInewsoul\fP(1). It creates a database of files, with meta\-data for MP3 and OGG Vorbis files, if Vorbis support was compiled in. There are two available shares databases, Normal and Buddies\-Only. Buddies\-Only is an optional shares database that, if enabled, is only accessible by the users you've chosen as "Buddies".
.LP 
Before running muscan, you will need a configured newsoul,  which can be done with \fImusetup\fP(1) or fI>musetup\-gtk\fP(1), and add some shared paths with either of the setup tools or with \fImuscan\fP(1).
.SH "OPTIONS"
.LP 
Muscan accepts the following options:
.TP 
.B \-c <\fIfilename\fP>, \-\-config <\fIfilename\fP>
Use a different config file.
.TP 
.B \-b, \-\-buddy
Select Buddy\-Only Shares.
.TP 
.B \-s <\fIdirectory\fP>, \-\-share <\fIdirectory\fP> 
Add directory to shares.
.TP 
.B \-u <\fIdirectory\fP>, \-\-unshare <\fIdirectory\fP> 
Remove directory from shares.
.TP 
.B \-r, \-\-rescan
Rescan shares.
.TP 
.B \-n, \-\-noscan
Do not rescan shares.
.TP 
.B \-v, \-\-verbose
Be verbose while scanning shares.
.TP 
.B \-l, \-\-list
Display list of shared directories.
.TP 
.B \-h, \-\-help
Display Help, version and exit.
.SH "EXAMPLE"
.LP 
To run a quick shares update, run:
.LP 
\fBmuscan\fP or \fBmuscan \-b\fP
.LP 
To rescan everything, run:
.LP 
\fBmuscan \-r\fP or \fBmuscan \-b \-r\fP
.LP 
Alternatively you use a different config file with:
.LP 
\fBmuscan \-\-config config.xml\fP
.LP 

.SH "FILES"
.TP 
 \fI~/.newsoul/config.xml\fR
The default location for the \fBnewsoul\fP config file.
.TP 
 \fI~/.newsoul/config.shares\fR
The default location for the Normal shares database.
.TP 
 \fI~/.newsoul/config.shares.state\fR
The default location for the active Normal shares database.
.TP 
 \fI~/.newsoul/config.buddyshares\fR
The default location for the Buddy shares database.
.TP 
 \fI~/.newsoul/config.buddyshares.state\fR
The default location for the active Buddy shares database's.
.SH "AUTHORS"
.LP 
Hyriand <hyriand@thegraveyard.org>
.LP 
daelstorm <daelstorm@gmail.com>
.SH "SEE ALSO"
.LP 
\fImucous\fP(1) \fImulog\fP(1) \fImurmur\fP(1) \fImuscand\fP(1) \fInscontrol\fP(1) \fInewsoul\fP(1) \fImuseeq\fP(1) \fImusetup\fP(1) \fImusetup\-gtk\fP(1)


CHANGELOG
This file tracks changes since the last Museek+ svn trunk revision.

-Switched from cmake to tup.
-Dropped PyMucipher bindings in favour of pycrypto (affects python bindings).
-Dropped Mucipher library in favour of nettle.
-muscan uses newsoul's configmanager (drops whole libxml++ dependency chain).
-Added systemd service file, but removed SysV one.
-Made some minor fixes.


src/muscan/muscan.cc
/* Tools - Tools for Museek (muscan)
 *
 * Copyright (C) 2003-2004 Hyriand <hyriand@thegraveyard.org>
 * Copyright 2008 little blue poney <lbponey@users.sourceforge.net>
 * Karol 'Kenji Takahashi' Woźniak © 2013
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "scanner.hh"
#include "../configmanager.h"
#include "../NewNet/nnlog.h"

#include <iostream>
#include <cstdlib>

using std::vector;
using std::string;
using std::cout;
using std::cerr;
using std::endl;
using std::map;

void help() {
    cout << "muscan [-c --config PATH] [-b --buddy] [-v --verbose] [-r --rescan] [-n --noscan] [-s --share PATH]... [-u --unshare PATH]..." << endl;
    cout << "or:" << endl;
    cout << "muscan [-c --config PATH] -l --list" << endl;
    cout << "Version 0.2.0" << endl;
    exit(-1);
}

int main(int argc, char **argv) {
    string config_file = string(getenv("HOME")) + "/.newsoul/config.xml";
    vector<string> add, remove;
    bool doList = false, rescan = false, noscan = false, doBuddy = false;

    for(int i = 1; i < argc; i++) {
        string arg = argv[i];
        if(arg == "-c" || arg == "--config") {
            ++i;
            if(i == argc)
                help();
            config_file = string(argv[i]);
        } else if(arg == "-s" || arg == "--share") {
            ++i;
            if(i == argc)
                help();
            add.push_back(string(argv[i]));
        } else if(arg == "-u" || arg == "--unshare") {
            ++i;
            if(i == argc)
                help();
            remove.push_back(string(argv[i]));
        } else if(arg == "-b" || arg == "--buddy") {
            doBuddy = true;
        } else if(arg == "-l" || arg == "--list") {
            doList = true;
        } else if(arg == "-v" || arg == "--verbose") {
            Scanner_Verbosity += 1;
        } else if(arg == "-r" || arg == "--rescan") {
            rescan = true;
        } else if(arg == "-n" || arg == "--noscan") {
            noscan = true;
        } else
            help();
    }
    
    if (Scanner_Verbosity >= 2){
        NNLOG.logEvent.connect(new NewNet::ConsoleOutput);
        NNLOG.enable("ALL");
    }
    
    Museek::ConfigManager config;
    config.load(config_file);
    //Muconf config(config_file);
    if(!config.hasDomain("shares") || !config.domain("shares").count("database")) {
        cerr << "config file '" << config_file << "' has incomplete or corrupt shares database" << endl;
        exit(-1);
    }
    if(!config.hasDomain("buddy.shares") || ! config.domain("buddy.shares").count("database")) {
        cerr << "config file '" << config_file << "' has incomplete or corrupt buddy shares database" << endl;
        exit(-1);
    }
    
    DirScanner root;
    string state;
    
    const string dbs("database");
    if (doBuddy) {
        string s = config.get("buddy.shares", dbs);
        state += s + ".state";
    } else {
        string s = config.get("shares", dbs);
        state += s + ".state";
    }
        
    root.load(state);

    if(doList) {
        map<string, DirEntry*>::iterator fit = root.folders.begin();
        for(; fit != root.folders.end(); ++fit)
            cout << (*fit).second->path << endl;
        
        return 0;
    }
    
    vector<string>::iterator it = add.begin();
    for(; it != add.end(); ++it)
        root.add(*it);

    it = remove.begin();
    for(; it != remove.end(); ++it) {
        map<string, DirEntry*>::iterator fit = root.folders.find(*it);
        if(fit != root.folders.end()) {
            delete (*fit).second;
            root.folders.erase(fit);
        }
    }

    if(rescan) {

        add.clear();

        map<string, DirEntry*>::iterator fit = root.folders.begin();
        for(; fit != root.folders.end(); ++fit)
            add.push_back((*fit).first);
        it = add.begin();
        
        root = DirScanner();
        
        for(; it != add.end(); ++it)
            root.add(*it);
    }
    if(! noscan)
        root.scan();
    root.save(state);
    DirScanner folded;
    root.fold(&folded);
        
    if (doBuddy) { 
        folded.save(config.get("buddy.shares", dbs));
    } else {
        folded.save(config.get("shares", dbs));
    }

    return 0;
}
